### PR TITLE
[Performance] SP×TP seq-sharded DSA indexer

### PR DESCRIPTION
**Pull request by mvasilijevicTT**
### What & why

Today `indexer_score` and `top-k` aren't TP-parallel — every TP chip runs them over the full query — and a **2-CCL logit all-reduce** (reduce-scatter + all-gather) is needed to sum the per-chip partial head-logits. This seq-shards the query over TP so both ops become TP-parallel (each chip works `S/(sp·tp)` rows), and replicates `wq_b` so each chip's head-sum is already complete on-chip — **dropping those 2 CCLs**. (Top-k's selected indices are gathered back over TP — one small CCL — so `sparse_sdpa` sees the same `[1,1,S/sp,k]` it always did.)

### Block-cyclic-aware indexer op

Under the block-cyclic KV layout a mid-slab chunk staggers each chip's query positions; a TP-sharded query must be scored *with* that stagger or top-k selects the wrong keys (a flat-geometry prototype regressed `test_sparse_mla_rotated` to 0.95, below the 0.98 gate). This extends `indexer_score`'s exact block-cyclic geometry to a 2D (SP×TP) sub-shard via a new `seq_subshard_axis` — rotation-exact by construction. Mechanics live in `device_causal_geometry` + the `indexer.py` block comment. **`mla.py` and the RoPE op are untouched.**

### Evidence

Correctness: `test_sparse_mla_rotated` 0.73→0.99 (gate 0.98); `accuracy` 12/12 incl. a 4×2 mesh. Measured on **LoudBox (8× Blackhole p150b, SP=2×TP=4 — 1/4-Galaxy proxy)**, sparse vs the merge-base baseline (TP-head-sharded indexer w/ the logit all-reduce); dense (v3.1 ring-MLA, indexer-independent) is the **control**.

**deepseek_v32** — critical-path device-kernel time (ms):
| scenario | sparse before | sparse after | Δ vs baseline | dense (control) |
|---|--:|--:|--:|--:|
| warm | 10.05 | **9.21** | **−8.3%** | 10.27 |
| cold | 95.06 | **86.79** | **−8.7%** | 72.64 |
| long | 32.31 | **19.34** | **−40.2%** | 76.98 |

**glm_5_1** — critical-path device-kernel time (ms):
| scenario | sparse before | sparse after | Δ vs baseline | dense (control) |
|---|--:|--:|--:|--:|
| warm | 9.07 | **7.87** | **−13.3%** | 3.71 |
| cold | 93.57 | **81.42** | **−13.0%** | 31.25 |
| long | 30.45 | **17.43** | **−42.8%** | 19.55 |

The win **grows with prefix** — both the removed all-reduce and the top-k work scale with `end_pos`. Per-op before/after + the code-verified dataflow graph: 📊 [interactive report](https://claude.ai/code/artifact/69360c1f-ffd2-42c1-9878-b03024805633).

### Notes for reviewers

- `seq_subshard_axis` is hashed (`has_value`) so a 2D call can't reuse an SP-only program's cached per-device offset.
- GLM (16 heads/chip) also pays a head→sequence reshard in `ttMLA._sparse_mla`; with this change GLM sparse now overtakes GLM dense at long prefix (17.43 vs 19.55).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mvasilijevic/indexer_sp32)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mvasilijevic/indexer_sp32)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:mvasilijevic/indexer_sp32)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mvasilijevic/indexer_sp32)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mvasilijevic/indexer_sp32)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mvasilijevic/indexer_sp32)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mvasilijevic/indexer_sp32)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mvasilijevic/indexer_sp32)

### CODE REVIEW

File: models/demos/deepseek_v3_d_p/tt/mla/indexer.py [hunk 1]
```diff
@@ -314,6 +320,23 @@ def _sp_all_gather(self, t, dim):
             cluster_axis=self.sp_axis,
         )
 
+    def _tp_all_gather(self, t, dim):
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tt/mla/indexer.py:324]: this should be a no op in the case of SPxSP sharded sparse attention right?

File: ttnn/cpp/ttnn/operations/experimental/indexer_score/indexer_score_nanobind.cpp [hunk 2]
```diff
@@ -90,6 +98,7 @@ void bind_indexer_score(nb::module_& mod) {
         nb::arg("cache_batch_idx") = std::nullopt,
         nb::arg("kv_len") = std::nullopt,
         nb::arg("cluster_axis") = std::nullopt,
+        nb::arg("seq_subshard_axis") = std::nullopt,
         nb::arg("block_cyclic_sp_axis") = std::nullopt,
```
Comment (ipotkonjak-tt) [inline ttnn/cpp/ttnn/operations/experimental/indexer_score/indexer_score_nanobind.cpp:105]: not a blocker for this PR, just a question

could we fold this into `seq_shard_axes=[axis_0, axis_1]`? Set to none the axis if sp is not sharded over it, or something similar

**Review by ipotkonjak-tt — APPROVED** [review]

File: models/demos/deepseek_v3_d_p/tt/mla/indexer.py [hunk 3]
```diff
@@ -563,46 +586,56 @@ def forward(
...
+            weights = ttnn.mesh_partition(weights, dim=2, cluster_axis=self.tp_axis)  # [1,H_idx,S/(sp·tp),1]
+            sq_local = seq_len // self.tp_factor
+            qc = 64 if sq_local % 64 == 0 else 32  # q_chunk must divide the per-chip query tile count
+        else:
+            qc = 64
+        # Causality is fused inside indexer_score (future columns -> -inf from chunk_start_idx), so no triu
+        # mask here. All H_idx heads are resident on-chip (wq_b replicated), so head_group_size=0 reads the
+        # key cache ONCE — but that needs L1 headroom, so k_chunk shrinks to 64 (vs 256 for a head-sharded
+        # slice).
+        cfg = ttnn.IndexerScoreProgramConfig(q_chunk_size=qc, k_chunk_size=min(64, end_pos), head_group_size=0)
```
Comment (skrsticTT) [inline models/demos/deepseek_v3_d_p/tt/mla/indexer.py]: it seams head_group_size=0 is now unconditional, we can have OOMs in some cases?

Comment (skrsticTT) [inline models/demos/deepseek_v3_d_p/tt/mla/indexer.py]: how we got 64 in the formula for k_chunk? it should depend from model to model i guess

**Review by skrsticTT — APPROVED** [review]
approved, address copilot's dealloc comments

File: models/demos/deepseek_v3_d_p/tt/mla/indexer.py (hunk 1 above)
Comment (mvasilijevicTT) [reply models/demos/deepseek_v3_d_p/tt/mla/indexer.py:324]: Yes. `_tp_all_gather` returns its input unchanged when `tp_factor == 1`. The only call site (line 705) is gated by `if tpsp:` where `tpsp = self._blockcyclic and self.tp_factor > 1` (line 622), so when `tp == 1` the call is skipped entirely — the identity guard inside the function is never even reached.

In the SP×TP path this PR adds (`tp > 1`), it does perform the real all-gather — the top-k was computed on the TP-split rows `[1,1,S/(sp·tp),k]` and has to be reassembled to the `[1,1,S/sp,k]` contract.

File: models/demos/deepseek_v3_d_p/tt/mla/indexer.py [hunk 4]
```diff
@@ -563,46 +586,56 @@ def forward(
...
+        # (block-cyclic-correct, cluster_axis=sp_axis), so every row already carries its true position; now
+        # split those rows over TP so each chip scores only S/(sp·tp) of them — indexer_score + topk shrink
+        # ~TP×. RoPE is per-row so the split is safe (no 2-D rope op needed). The score is told the TP axis via
+        # seq_subshard_axis below, so its EXACT block-cyclic geometry adds each device's tp_rank*Sq' sub-offset
+        # (rotation-safe). topk runs on the sub-rows; indices are all-gathered back over TP to the [1,1,S/sp,k]
+        # contract so mla.py / sparse_sdpa are unchanged (both DeepSeek and GLM ride this one path).
+        tpsp = self._blockcyclic and self.tp_factor > 1
+        if tpsp:
+            q_dev = ttnn.mesh_partition(q_dev, dim=2, cluster_axis=self.tp_axis)  # [1,H_idx,S/(sp·tp),D_idx]
+            weights = ttnn.mesh_partition(weights, dim=2, cluster_axis=self.tp_axis)  # [1,H_idx,S/(sp·tp),1]
```
[the remark this answers is not in the document]
Comment (mvasilijevicTT) [reply models/demos/deepseek_v3_d_p/tt/mla/indexer.py:630]: Fixed in f5ad2efd410 — capture the full-S `q_dev`/`weights` before `mesh_partition` and `ttnn.deallocate` them once the TP-split replacements exist (matches the mla.py mesh-partition dealloc convention).

File: models/demos/deepseek_v3_d_p/tt/mla/indexer.py [hunk 5]
```diff
@@ -657,9 +688,15 @@ def forward(
...
         topk_valid_length = end_pos if self._blockcyclic else None
-        return ttnn.experimental.topk_large_indices(
+        idx = ttnn.experimental.topk_large_indices(
             logits, k=min(self.index_args.index_topk, end_pos), valid_length=topk_valid_length
         )
+        # TP×SP: topk ran on the TP-seq-sharded rows ([1,1,S/(sp·tp),k]); regather over TP back to the
+        # [1,1,S/sp,k] contract so sparse_sdpa/mla.py are unchanged. (Redundant TP-round-trip for GLM's
+        # head→seq reshard, which re-splits it; correct regardless. tp=1: no-op.)
+        if tpsp:
+            idx = self._tp_all_gather(idx, dim=2)
```
[the remark this answers is not in the document]
Comment (mvasilijevicTT) [reply models/demos/deepseek_v3_d_p/tt/mla/indexer.py:707]: Fixed in f5ad2efd410 — `ttnn.deallocate` the TP-sharded top-k after the TP all-gather regathers it to `[1,1,S/sp,k]`.

File: ttnn/cpp/ttnn/operations/experimental/indexer_score/indexer_score_nanobind.cpp [hunk 6]
```diff
@@ -62,6 +62,14 @@ void bind_indexer_score(nb::module_& mod) {
...
                 index along this axis (Sq = q seq len). None = linear device order
                 (1 on a single device, so chunk_start_idx is used as-is).
+            seq_subshard_axis: optional int, the SECOND mesh axis (TP) the query seq is
+                block-cyclically sub-sharded over, on top of the SP block-cyclic layout.
+                Set it alongside a named cluster_axis + block_cyclic when the query rows
+                are further split across TP (block_cyclic_chunk_local == tp*q_isl): the
+                exact geometry adds tp_rank*Sq to each device's block-cyclic position, so
+                the causal mask stays correct under rotated (mid-slab) starts — unlike the
+                cluster_axis=None flat path, which is linear-approximate there. Requires a
+                block-cyclic layout and an axis distinct from cluster_axis.
```
[the remark this answers is not in the document]
Comment (mvasilijevicTT) [reply ttnn/cpp/ttnn/operations/experimental/indexer_score/indexer_score_nanobind.cpp:72]: Fixed in f5ad2efd410 — reworded the `block_cyclic_chunk_local` paragraph so it no longer contradicts the `seq_subshard_axis` path: the `tp*q_isl` case with a named `cluster_axis` now correctly says it requires `seq_subshard_axis` (matching the TT_FATAL in indexer_score_device_operation.cpp).

File: models/demos/deepseek_v3_d_p/tt/mla/indexer.py (hunk 3 above)
Comment (mvasilijevicTT) [reply models/demos/deepseek_v3_d_p/tt/mla/indexer.py]: Yes. Basically, you have to choose k_chunk size so that it fits memory (up to 96 for deep seek, 256 for GLM).  Head group = 0 is noticeably faster than 64, so I'd say !=0 is not an option for perf reasons.

Comment (mvasilijevicTT) [reply models/demos/deepseek_v3_d_p/tt/mla/indexer.py]: Good catch! 64 gives the best performance for DeepSeek. And you are right, for GLM this should be 224. I'll make it configurable.
